src/docker/shared.rs

```diff
@@ -763,6 +763,164 @@ mod tests {
+                None,
+                msg_info,
+            )?
+            .ok_or_else(|| eyre::eyre!("could not find cross workspace"))?;
+
+            metadata.workspace_root = get_workspace_root(subdir)?;
+            metadata.target_directory = metadata.workspace_root.join("target");
+
+            Ok(metadata)
+        }
+
+        fn home() -> Result<PathBuf> {
+            home::home_dir().ok_or_else(|| eyre::eyre!("could not find home directory"))
+        }
+
+        fn get_cwd() -> Result<PathBuf> {
+            // we need this directory to exist for Windows
+            let path = temp::dir()?.join("Documents").join("package");
+            fs::create_dir_all(&path)?;
+            Ok(path)
+        }
+
+        fn get_sysroot() -> Result<PathBuf> {
+            Ok(home()?
+                .join(".rustup")
+                .join("toolchains")
+                .join("stable-x86_64-unknown-linux-gnu"))
+        }
+
+        fn get_directories(subdir: bool, mount_finder: &MountFinder) -> Result<Directories> {
+            let msg_info = MessageInfo::default();
+            let metadata = cargo_metadata(subdir, msg_info)?;
+            let cwd = get_cwd()?;
+            let sysroot = get_sysroot()?;
+            Directories::create(mount_finder, &metadata, &cwd, &sysroot)
+        }
+
+        fn path_to_posix(path: &Path) -> Result<String> {
+            #[cfg(target_os = "windows")]
+            {
+                path.as_wslpath()
+            }
+            #[cfg(not(target_os = "windows"))]
+            {
+                path.as_posix()
+            }
+        }
+
+        #[track_caller]
+        fn paths_equal(x: &Path, y: &Path) -> Result<()> {
+            assert_eq!(path_to_posix(x)?, path_to_posix(y)?);
+            Ok(())
+        }
+
+        #[test]
+        fn test_host() -> Result<()> {
+            let vars = unset_env();
+            let mount_finder = MountFinder::new(vec![]);
+            let directories = get_directories(false, &mount_finder)?;
+            paths_equal(&directories.cargo, &home()?.join(".cargo"))?;
+            paths_equal(&directories.xargo, &home()?.join(".xargo"))?;
+            paths_equal(&directories.host_root, &get_workspace_root(false)?)?;
+            assert_eq!(
+                &directories.mount_root,
+                &path_to_posix(&get_workspace_root(false)?)?
+            );
+            assert_eq!(&directories.mount_cwd, &path_to_posix(&get_cwd()?)?);
+
+            reset_env(vars);
+            Ok(())
+        }
+
+        #[test]
+        fn test_docker_in_docker() -> Result<()> {
+            if !cfg!(target_os = "linux") {
+                return Ok(());
+            }
+
+            let vars = unset_env();
+
+            let engine = create_engine(MessageInfo::default());
+            let hostname = env::var("HOSTNAME");
+            if engine.is_err() || hostname.is_err() {
+                reset_env(vars);
+                return Ok(());
```

Comment:
```suggestion
                reset_env(vars);
                eprintln!("no hostname found");
                return Ok(());
```

Reply:
I changed it to mention that either the container engine or hostname isn't defined. It's not very likely `HOSTNAME` won't be defined.